### DISCUSSION TITLE
backpack table users

### DISCUSSION THREAD

**Discussion by soluzione-software** [Laravel-Backpack/community-forum#437] [Q&A (Help)] [answered]
I would like to differentiate backpack users from front end users.
leave the front end users in the users table while the backend ones in a new admin table. how can I do?

**Comment by promatik** [comment]
Hi @soluzione-software!

There is many approaches you can take here.

**1) Different guard names**

Usually developers keep the users in the same table, using the same `App\Models\User`, but using a different guard name.
Backpack uses by default the guard name `'backpack'`, you can check this configuration at `config\backpack\base.php` » `guard`.
You can have a different guard name, usually `web` (it's the default on Laravel) for the web users.
Everything works the same way, login, regist and so on, but when setting the roles and permissions, you set the different guard name. Web users will never be able to see the admin panel.

**2) Separate users in many tables**
Besides different guard names, you can separate users in different tables. For this you must create a new user Model, and use different models for admins and web users.
Backpack allows you to set the user model class at `config\backpack\base.php` » `user_model_fqn`, so you can either change Backpack users to `App\Model\Admin::class`, or create a new one for the web users.
I would create a new one for the Admin, and that new `Admin::class` would extend the `User::class`. Changing just what is needed, like the `$table = 'admins'`.

Also, take a look at `configs\auth.php`, you may want to organize your guards and providers there.

**Comment by soluzione-software** [comment]
Thank you for your answer.
I have another question; I am using [Laravel-Backpack/PermissionManager] and after doing the separate of table the method user()->can('...') seems to return false every time, why?

**Comment by promatik** [comment, marked as answer]
@soluzione-software you have more information about multiple guards on spatie laravel-permission docs: https://spatie.be/docs/laravel-permission/v5/basic-usage/multiple-guards 👌

Make sure about;
1) your `user()->can('...')` is being applied to the correct guard, if you pass nothing, it will be applied to the default guard. If you want to specify a guard, send `user()->can('...', $guard)`.

2) your roles/guards are being correctly assigned on `roles` table.
![image](https://user-images.githubusercontent.com/1838187/225646288-e85a36b6-6a9e-4025-bd3b-00357840f35c.png)


3) your relations between users and roles/permissions are being correctly stored on `model_has_roles` and `model_has_permissions`.
![image](https://user-images.githubusercontent.com/1838187/225646549-9f3f98b4-f864-4d3b-920c-a9d93234606b.png)
Note there must be some `App\Models\User` and some `App\Models\Admin` 👌

One of this places may be wrong, and that's why user->can is not working, let me know 🙌

**Comment by soluzione-software** [comment]
i try to explain better:
my config.auth:
```php
'defaults' => [
	'guard' => 'web',
	'passwords' => 'users',
],
'guards' => [
      'web' => [
          'driver' => 'session',
          'provider' => 'users',
      ],
      'admin' => [
          'driver' => 'session',
          'provider' => 'admins',
      ],
  ],
'providers' => [
      'users' => [
          'driver' => 'eloquent',
          'model' => App\Models\User::class,
      ],
      'admins' => [
          'driver' => 'eloquent',
          'model' => App\Models\Admin::class,
      ],
  ],
'passwords' => [
      'users' => [
          'provider' => 'users',
          'table' => 'password_resets',
          'expire' => 60,
          'throttle' => 60,
      ],
      'admins' => [
          'provider' => 'admins',
          'table' => 'password_resets',
          'expire' => 60,
          'throttle' => 60,
      ],
  ],
```

my config.backpack.base
```php
'user_model_fqn' => config('auth.providers.admins.model'),
...
'guard' => 'backpack',
'passwords' => 'backpack',
...
```

my config.backpack.permissionmanager
```php
'models' => [
    'user'       => config('backpack.base.user_model_fqn', \App\Models\User::class),
    'permission' => Backpack\PermissionManager\app\Models\Permission::class,
    'role'       => Backpack\PermissionManager\app\Models\Role::class,
],
```

my Models\Admin
```php
<?php

namespace App\Models;

use Backpack\CRUD\app\Models\Traits\CrudTrait;
use Illuminate\Contracts\Auth\MustVerifyEmail;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Laravel\Sanctum\HasApiTokens;
use Spatie\Permission\Models\Role;
use Spatie\Permission\Traits\HasRoles;

class Admin extends User
{
    use HasRoles;
    protected $table = 'admins';
}
```

but Auth::user() is null, backpack_user() is correct

thanks

**Reply by promatik** [reply]
Ok, so Admin part is working from what I understood ...
How are you doing the login of your web users?
Make sure those users have the correct guard `web`.

**Comment by soluzione-software** [comment]
ok i solve, i check configuration and backpack set the 'backpack' guard and not web
i make override to permissionmanager to use this guard, so backpack guard is backpack and front end guard is web
it seems to work
thanks
